### PR TITLE
Puppeteer -> Playwright

### LINKED ISSUE #350: 🎭Change to Playwright, support all the browsers
https://github.com/microsoft/playwright
https://twitter.com/aslushnikov/status/1220045633282662401

### LINKED ISSUE #354: ✅npx qawolf test --all-browsers

### LINKED ISSUE #356: 🧺npx qawolf bitbucket

### LINKED ISSUE #365: ⬆️Switch back to upstream ffmpeg-static
https://github.com/eugeneware/ffmpeg-static/commit/192413704dda43a9bfb64bb4d16317b849fd4961

### PR DESCRIPTION
**Pull request by jperl**
Resolves #350 
Resolves #354
Resolves #356 
Resolves #365

To Do:

- [x] get types to build
- [x] QAWolfBrowser -> QAWolfBrowserContext, include reference to Browser
- [x] refactor pages
- [x] rename generated code from `browser = await launch -> context =...`
- [x] fix all tests for chrome
- [x] add test that recorder works on a new page
- [x] fix all tests for firefox
- [x] fix all tests for webkit
- [x] npx qawolf test [--all-browsers, --chromium, --webkit, --firefox], document it #354
- [x] add connect support, add it to docs
- [x] fix create
- [x] update playwright version
- [x] confirm recording works on ci as expected
- [x] publish alpha
- [x] include browser type in artifact path
- [x] update all ci templates with necessary dependencies
- [x] add bitbucket cli command
- [ ] fix tests on windows

Remaining playwright issues:

- https://github.com/microsoft/playwright/issues/657
- https://github.com/microsoft/playwright/issues/808

### CODE REVIEW

**Comment by jperl** [comment]
Todo:

- [ ] docs Device -> DeviceDescriptor
- [ ] submit issue for DeviceDescriptor types

File: packages/web-tests/tests/Recorder.test.ts [hunk 1]
```diff
@@ -56,7 +56,7 @@ describe("Recorder", () => {
 
     const page = await browser.page();
 
-    // from https://github.com/GoogleChrome/puppeteer/issues/4119#issue-417279184
+    // from https://github.com/GoogleChrome/playwright/issues/4119#issue-417279184
```
Comment (odinho) [inline packages/web-tests/tests/Recorder.test.ts]: I bet you didn't meant to do this ;)

**Comment by jperl** [comment]
Need to rethink Browser vs BrowserContext, and how to support .connect

**Comment by jperl** [comment]
Based on discussion with @flaurida, rename browser flags to branded names: `--chromium -> --chrome`, `--webkit -> --safari`

File: docs/docs/faq.md [hunk 2]
```diff
@@ -13,26 +13,26 @@ We found the open source recorders brittle and limited in their functionality. T
 
 We are philosophically opposed to closed-source test recorders and proprietary test frameworks. We believe that if you put in the work to create your tests, you should own them, be able to edit them as you like, and be able to run them on your own infrastructure.
 
-## Why Puppeteer over Selenium?
+## Why Playwright over Selenium?
 
-We found Puppeteer to be more stable than Selenium in practice and we're not [the](https://medium.com/coursera-engineering/improving-end-to-end-testing-at-coursera-using-puppeteer-and-jest-5f1bac9cd176) [only](https://news.ycombinator.com/item?id=20505711) [ones](https://news.ycombinator.com/item?id=20506053).
+We found Puppeteer (the source of Playwright) to be more stable than Selenium in practice and we're not [the](https://medium.com/coursera-engineering/improving-end-to-end-testing-at-coursera-using-puppeteer-and-jest-5f1bac9cd176) [only](https://news.ycombinator.com/item?id=20505711) [ones](https://news.ycombinator.com/item?id=20506053).
```
Comment (flaurida) [inline docs/docs/faq.md]: the precursor to Playwright?

File: packages/browser/src/find/findProperty.ts [hunk 3]
```diff
@@ -11,7 +11,7 @@ export const findProperty = async (
   const element = await find(page, selector, options);
 
   const propertyHandle = await element.getProperty(property);
-  const propertyValue = await propertyHandle.jsonValue();
+  const propertyValue = await propertyHandle!.jsonValue();
```
Comment (flaurida) [inline packages/browser/src/find/findProperty.ts]: any change would be `null`? why `!`?

File: packages/browser/src/keyboard/serializeStrokes.ts [hunk 4]
```diff
@@ -1,12 +1,11 @@
...
 import { KeyEvent } from "@qawolf/types";
 import { uniq } from "lodash";
-import KeyDefinitions from "puppeteer/lib/USKeyboardLayout";
+import { keyDefinitions } from "playwright-core/lib/usKeyboardLayout";
 import { Stroke } from "./Stroke";
-import "../USKeyboardLayoutTypes";
 
 const USKeys = uniq(
-  Object.values(KeyDefinitions).map(definition => definition.key)
+  Object.values(keyDefinitions).map((definition: any) => definition.key)
```
Comment (flaurida) [inline packages/browser/src/keyboard/serializeStrokes.ts]: can you add a type? even if it's like `{ definition: string }` etc

File: packages/browser/src/find/findProperty.ts (hunk 3 above)
Comment (jperl) [reply packages/browser/src/find/findProperty.ts]: Property handle could be null. I guess I should do this instead?

```
const propertyHandle = await element.getProperty(property);
if (!propertyHandle) return null;
```

File: packages/browser/src/keyboard/serializeStrokes.ts (hunk 4 above)
Comment (jperl) [reply packages/browser/src/keyboard/serializeStrokes.ts]: Good catch. I don't need any, now that playwright fixed their types.

File: packages/browser/src/page/findPage.ts [hunk 5]
```diff
@@ -1,63 +1,68 @@
...
 
+  // TODO waiting for https://github.com/microsoft/playwright/issues/657
   // when headless = false the tab needs to be activated
   // for the execution context to run
-  await page.bringToFront();
+  // await page.bringToFront();
 
   if (options.waitForRequests) {
-    await page.qawolf.waitForRequests();
+    await page.waitForRequest(/.*/g);
```
Comment (jperl) [inline packages/browser/src/page/findPage.ts]: this should respect the timeout specified

**Comment by jperl** [comment]
The only remaining task is to fix windows tests. From what I can tell, they are all fixed except for those broken due to this bug with playwright https://github.com/microsoft/playwright/issues/808

I am going to go ahead and merge this PR. However we hold off releasing until playwright fixes ^ and our windows tests pass.

**Comment by jperl** [comment]
Fix windows tests blocked by https://github.com/isaacs/rimraf/pull/209

**Pull request merged**